spec.md

```diff
@@ -683,21 +703,28 @@ message ControllerPublishVolumeRequest {
 }
 
 message ControllerPublishVolumeResponse {
-  message Result {
-    // The SP specific information that will be passed to the Plugin in
-    // the subsequent `NodePublishVolume` call for the given volume.
-    // This information is opaque to the CO. This field is OPTIONAL.
-    map<string, string> publish_volume_info = 1;
-  }
-
-  // One of the following fields MUST be specified.
-  oneof reply {
-    Result result = 1;
-    Error error = 2;
-  }
+  // The SP specific information that will be passed to the Plugin in
+  // the subsequent `NodePublishVolume` call for the given volume.
+  // This information is opaque to the CO. This field is OPTIONAL.
+  map<string, string> publish_volume_info = 1;
 }
 ```
 
+##### ControllerPublishVolume Errors
+
+If the plugin is unable to complete the ControllerPublishVolume call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+Condition | gRPC Code | Description | Recovery Behavior
+| --- | --- | --- | --- |
+| Volume does not exists | 5 NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |
+| Node does not exists | 5 NOT_FOUND | Indicates that a node corresponding to the specified `node_id` does not exist. | Caller MUST verify that the `node_id` is correct and that the node is available and has not been terminated or deleted before retrying with exponential backoff. |
+| Volume published to another node | 6 ALREADY_EXISTS | Indicates that a volume corresponding to the specified `volume_id` is already attached to another node and does not support multi-node attach. If this error code is returned, the Plugin SHOULD specify the `node_id` of the node the volume is already attached to as part of the gRPC `status.message`. | Caller SHOULD ensure the specified volume is not attached to any other node before retrying with exponential back off. |
+| Max volumes attached | 8 RESOURCE_EXHAUSTED | Indicates that the maximum supported number of volumes that can be attached to the specified node are already attached. Therefore, this operation will fail until at least one of the existing attached volumes is detached from the node. | Caller MUST ensure that the number of volumes already attached to the node is less then the maximum supported number of volumes before retrying with exponential backoff. |
```

Comment:
RESOURCE_EXHAUSTED is also returned for other reasons by the grpc library: https://github.com/grpc/grpc/blob/master/doc/statuscodes.md

Reply:
suggest FAILED_PRECONDITION instead